src/Xamarin.Android.Build.Tasks/Utilities/MonoAndroidHelper.cs

```diff
@@ -623,5 +624,49 @@ public static string TryGetAndroidJarPath (TaskLoggingHelper log, string platfor
 			}
 			return result;
 		}
+
+		public static string ExtractNdkResources (string toolDirectory)
+		{
+			if (!OS.IsMac)
+				return toolDirectory;
+
+			var embeddedNdkTools = new string [] {
+				"aarch64-linux-android-as",
+				"aarch64-linux-android-ld",
+				"aarch64-linux-android-strip",
+				"arm-linux-androideabi-as",
+				"arm-linux-androideabi-ld",
+				"arm-linux-androideabi-strip",
+				"i686-linux-android-as",
+				"i686-linux-android-ld",
+				"i686-linux-android-strip",
+				"x86_64-linux-android-as",
+				"x86_64-linux-android-ld",
+				"x86_64-linux-android-strip",
+			};
+
+			var home = Environment.GetFolderPath (Environment.SpecialFolder.Personal);
+			var ndkToolRoot = Path.Combine (home, "Library", "Developer", "Xamarin", "ndk-tools");
+			var assem = Assembly.GetExecutingAssembly ();
+
+			foreach (var ndkTool in embeddedNdkTools) {
+				var ndkToolDestination = Path.Combine (ndkToolRoot, ndktool);
+				if (!File.Exists (ndkToolDestination)) {
```

Comment:
Problem here will be "cache invalidation": when Xamarin.Android *10.1* comes out, bundling tools from NDK r20 (or whatever), how do we know that we need to *replace* the files within `ndkToolDestination` with the embedded resources?

Possible solution: we set the timestamp of the files within `ndkToolDestination` to have the same file timestamp as `Assembly.GetExecutingAssembly()`.

Slight problem is that when 10.0.1 comes out -- which we'll assume is still NDK r19, as d16-3 uses NDK r19 -- we'd overwrite these files with "identical" values.  We might want to *also* use `Files.CopyIfChanged()`?

Relatedly: how long will this method take to execute?  `Files.CopyIfChanged()` might be moot, and we really just need the file timestamp check.

Reply:
Agreed, this is just a quick prototype to unblock other Notarization related efforts. If we do ultimately choose to go this route we'll absolutely need to address these concerns, and ensure that the occasional "one time" performance hit we will incur will be acceptable.